### PR TITLE
Apply suggested fixes from ShellCheck in sh scripts

### PR DESCRIPTION
**Pull request by ishanjalan**
https://github.com/koalaman/shellcheck/

### CODE REVIEW

File: src/SMAPI.Installer/assets/unix-launcher.sh [hunk 1]
```diff
@@ -102,37 +102,39 @@ else
...
                 gnome-terminal)
                     # consumes all arguments after --
-                    exec $TERMINAL_NAME -- env TERM=xterm $LAUNCH_FILE "$@"
+                    exec "$TERMINAL_NAME" -- env TERM=xterm $LAUNCH_FILE "$*"
                     ;;
 
                 kitty)
                     # consumes all trailing arguments
-                    exec $TERMINAL_NAME env TERM=xterm $LAUNCH_FILE "$@"
+                    exec "$TERMINAL_NAME" env TERM=xterm $LAUNCH_FILE "$*"
```
Comment (jltaylor-us) [inline src/SMAPI.Installer/assets/unix-launcher.sh]: changing from `"$@"` to `"$*"` is almost certainly incorrect, but as long as no one ever puts spaces inside the an argument then no one will notice.

(The mac script generation is broken in a similar way, but it was already broken.  The right way to escape things going into the new script is with the `%q` format specifier to the bash `printf` builtin.)

File: build/unix/prepare-install-package.sh [hunk 2]
```diff
@@ -25,14 +25,14 @@ declare -A msBuildPlatformNames=(["linux"]="Unix" ["macOS"]="OSX" ["windows"]="W
...
-    read version
+    read -r version
 fi
 
 
 ##########
 ## Move to SMAPI root
 ##########
-cd "`dirname "$0"`/../.."
+cd "$(dirname "$0")/../.." || exit
```
Comment (jltaylor-us) [inline build/unix/prepare-install-package.sh:35]: Exiting in an error condition with no error message isn't great; exiting with a "success" exit code is worse.

It doesn't look like these files do any other sort of error checking to make sure commands succeeded.  If you want to add return status checking then `set -e` might be a better approach.

File: build/unix/prepare-install-package.sh [hunk 3]
```diff
@@ -66,10 +66,10 @@ for folder in ${folders[@]}; do
     echo ""
     echo ""
 
-    for modName in ${bundleModNames[@]}; do
+    for modName in "${bundleModNames[@]}"; do
         echo "Compiling $modName for $folder..."
         echo "-------------------------------------------------"
-        dotnet publish src/SMAPI.Mods.$modName --configuration $buildConfig -v minimal --runtime "$runtime" -p:OS="$msbuildPlatformName" -p:GamePath="$gamePath" -p:CopyToGameFolder="false"
+        dotnet publish src/SMAPI.Mods."$modName" --configuration $buildConfig -v minimal --runtime "$runtime" -p:OS="$msbuildPlatformName" -p:GamePath="$gamePath" -p:CopyToGameFolder="false"
```
Comment (Pathoschild) [inline build/unix/prepare-install-package.sh]: Shouldn't we just quote the whole argument?
```suggestion
        dotnet publish "src/SMAPI.Mods.$modName" --configuration $buildConfig -v minimal --runtime "$runtime" -p:OS="$msbuildPlatformName" -p:GamePath="$gamePath" -p:CopyToGameFolder="false"
```

File: build/unix/prepare-install-package.sh [hunk 4]
```diff
@@ -117,13 +117,13 @@ for folder in ${folders[@]}; do
...
     # installer DLL config
-    if [ $folder == "windows" ]; then
+    if [ "$folder" == "windows" ]; then
         cp "$installAssets/windows-exe-config.xml" "$packagePath/internal/windows/install.exe.config"
     fi
 
     # bundle root files
     for name in "StardewModdingAPI" "StardewModdingAPI.dll" "StardewModdingAPI.pdb" "StardewModdingAPI.xml" "steam_appid.txt"; do
-        if [ $name == "StardewModdingAPI" ] && [ $folder == "windows" ]; then
+        if [ $name == "StardewModdingAPI" ] && [ "$folder" == "windows" ]; then
```
Comment (Pathoschild) [inline build/unix/prepare-install-package.sh]: Shouldn't `$name` be quoted too while we're at it?
```suggestion
        if [ "$name" == "StardewModdingAPI" ] && [ "$folder" == "windows" ]; then
```

File: build/unix/prepare-install-package.sh [hunk 5]
```diff
@@ -181,15 +181,16 @@ done
...
     # disable developer mode in main package
     sed --in-place --expression="s/\"DeveloperMode\": true/\"DeveloperMode\": false/" "$packagePath/internal/$folder/bundle/smapi-internal/config.json"
 
     # convert bundle folder into final 'install.dat' files
     for path in "$packagePath/internal/$folder" "$packageDevPath/internal/$folder"; do
-        pushd "$path/bundle" > /dev/null
-        zip "install.dat" * --recurse-paths --quiet
-        popd > /dev/null
+        pushd "$path/bundle" > /dev/null || exit
+        #https://github.com/koalaman/shellcheck/wiki/SC2035 this should be either ./* or -- *
```
Comment (Pathoschild) [inline build/unix/prepare-install-package.sh]: Unneeded comment:
```suggestion
```

File: src/SMAPI.Installer/assets/install on Linux.sh [hunk 6]
```diff
@@ -1,4 +1,4 @@
 #!/bin/bash
 
-cd "`dirname "$0"`"
+cd "$(dirname "$0")" || exit
```
Comment (Pathoschild) [inline src/SMAPI.Installer/assets/install on Linux.sh:3]: If we're going to handle all these possible error edge cases, we should log an error message indicating what failed too.

**Review by Pathoschild — CHANGES_REQUESTED** [review]

File: src/SMAPI.Installer/assets/unix-launcher.sh (hunk 1 above)
Comment (Pathoschild) [reply src/SMAPI.Installer/assets/unix-launcher.sh:137]: Spaces in Steam command-line arguments are certainly possible.

**Comment by Pathoschild** [comment]
List of included fixes per discussion with @ishanjalan:

level | suggested fix
---- | -------------
error | [Argument mixes string and array. Use * or separate argument.](https://github.com/koalaman/shellcheck/wiki/SC2145)
error | [Double quote array expansions to avoid re-splitting elements.](https://github.com/koalaman/shellcheck/wiki/SC2068)
warn | [Declare and assign separately to avoid masking return values.](https://github.com/koalaman/shellcheck/wiki/SC2155)
warn | [Use `... \|\| exit` or `... \|\| return` in case command fails.](https://github.com/koalaman/shellcheck/wiki/SC2164)
warn | [Quote to prevent word splitting.](https://github.com/koalaman/shellcheck/wiki/SC2046)
hint | [read without -r will mangle backslashes.](https://github.com/koalaman/shellcheck/wiki/SC2162)
hint | [Use `$(...)` notation instead of legacy backticked \`...\`.](https://github.com/koalaman/shellcheck/wiki/SC2006)
hint | [Double quote to prevent globbing and word splitting.](https://github.com/koalaman/shellcheck/wiki/SC2086)
hint | [Use `./*glob*` or `-- *glob*` so names with dashes won't become options.](https://github.com/koalaman/shellcheck/wiki/SC2035)

File: src/SMAPI.Installer/assets/unix-launcher.sh (hunk 1 above)
Comment (Pathoschild) [reply src/SMAPI.Installer/assets/unix-launcher.sh]: Spaces in Steam command-line arguments are certainly possible.

Based on [the wiki page for the suggested fix](https://github.com/koalaman/shellcheck/wiki/SC2145), `$@` seems to be correct for the way we're using it. We're not appending it to a word, so the quirks don't apply in this case.

**Comment by Pathoschild** [comment]
@ishanjalan I'll close this PR since it changes too many different things at once, and it still has some unresolved comments. Feel free to submit smaller PRs that change specific things so they can be discussed separately. Sorry for the inconvenience!

**Pull request closed without merging**